server/aesdsocket.c

```diff
@@ -15,42 +15,245 @@
+ *
+ */
+static void *timer_handler(void *signalno) {
+
+	while (1)
+	{
+		char timestampBuffer[100];
+
+		// Get current time
+		time_t currentTime = time(NULL);
+
+		// Convert time to local time
+		struct tm * time_structure = localtime(&currentTime);
+		if (time_structure == NULL) {
+			perror("convertion of time to localtime failed");
+			exit(EXIT_FAILURE);
+		}
+
+		// Format the timestamp
+		int timestampLength = strftime(timestampBuffer, sizeof(timestampBuffer), "timestamp:%y %b %d	%k:%M:%S\n", time_structure);
+		if (timestampLength == 0) {
+			perror("Formating of timestamp failed");
+			exit(EXIT_FAILURE);
+		}
+
+		// printing timestamp before dumping in file
+		printf("timestamp: %s\n", timestampBuffer);
+
+		/* Appending timestamp onto the file */
+
+		// Open or create the data file for writing, and append data
+		int data_file = open(DATA_FILE, O_WRONLY | O_CREAT | O_APPEND, FILE_PERMISSIONS);
+		if (data_file == -1) {
+			perror("File open failed"); // Print an error message if file open fails
+			syslog(LOG_ERR, "File open failed: %m"); // Log file open error
+			exit(EXIT_FAILURE);
+		}
+
+		// Obtain the mutex lock to protect shared resources
+		if (pthread_mutex_lock(&mutex_lock) != 0) {
+			exit(EXIT_FAILURE);
+		}
+
+		// Write the timestamp to the data file
+		int writeStatus = write(data_file, timestampBuffer, timestampLength);
+
+		// Release the mutex lock
+		if (pthread_mutex_unlock(&mutex_lock) != 0) {
+			exit(EXIT_FAILURE);
+		}
+
+		if (writeStatus == -1) {
+			printf("Error writing to the file");
+			exit(EXIT_FAILURE);
+		}
+
+		// Close the data file
+		close(data_file);
+
+		// The string should be appended to the /var/tmp/aesdsocketdata file every 10 seconds.
+		// Sleep for 10 seconds before the next iteration
+		sleep(10);
+	}
+
+	// Thread completed successfully
+	pthread_exit(NULL);
+}
+
+/**
+ * @name: thread_func
+ * @brief: Thread handler function for receiving and sending data
+ * @params: void *thread_parameters: thread parameters
+ * @return: void *thread_parameters
+ */
+void* thread_func(void *thread_param) {
+
+    struct thread_data *thread_func_args = (struct thread_data *)thread_param;
+
+    char *buffer = (char *)malloc(sizeof(char) * BUFFER_SIZE);
+    if (buffer == NULL) {
+        syslog(LOG_ERR, "Memory allocation failed"); // Log memory allocation error
+        exit(EXIT_FAILURE); // Return an error status
+    }
+    memset(buffer, 0, BUFFER_SIZE);
+
+	// Receive and append data
+    int data_file = open(DATA_FILE, O_WRONLY | O_CREAT | O_APPEND, FILE_PERMISSIONS);
+    if (data_file == -1) {
+        perror("File open failed"); // Print an error message if file open fails
+		syslog(LOG_ERR, "File open failed: %m"); // Log file open error
+        free(buffer);
+        exit(EXIT_FAILURE);
     }
+
+	ssize_t bytes_received;
+    while ((bytes_received = recv(thread_func_args->client_socket, buffer, BUFFER_SIZE, 0)) > 0) {
+        write(data_file, buffer, bytes_received);
```

Comment:
@sane5805 how would this handle a case where packet length was greater than `BUFFER_SIZE`?  Would it be possible that a partial packet would be interleaved by timestamp prints or prints from another packet?

Reply:
Also missing write locking here.